subsys/testsuite/ztest/include/ztest_test.h

```diff
@@ -27,6 +29,62 @@ struct unit_test {
 	uint32_t thread_options;
 };
 
+/**
+ * A single node of test suite. Each node should be added to unit_test_head to
+ * form a linked list that can be run via ztest_run_registered_test_suites().
+ */
+struct unit_test_node {
+	/** The name of the test suite. */
+	const char *name;
+	/** Pointer to the test suite. */
+	struct unit_test *suite;
+	/**
+	 * An optional pragma function to determine if the test should run. If NULL, then the test
+	 * will only run once on the first attempt.
+	 */
+	bool (*pragma)(const void *state);
+	/** The next suite to try and run. */
+	struct unit_test_node *next;
+};
+
+/** The head of the registered unit test suites. */
+extern struct unit_test_node *z_ztest_unit_test_head;
+
+/**
+ * Create and register a ztest suite. Using this macro creates a new test suite (using
+ * ztest_test_suite). It then creates a struct unit_test_node and an initializer function which will
+ * be run via a SYS_INIT call at the APPLICATION initialization level and priority 99.
+ *
+ * Tests can then be run by calling ztest_run_registered_test_suites(const void *state) by passing
+ * in the current state. See the documentation for ztest_run_registered_test_suites for more info.
+ *
+ * @param SUITE_NAME The name of the suite (see ztest_test_suite for more info)
+ * @param PRAGMA A function to test against the state and determine if the test should run.
+ */
+#define ztest_register_test_suite(SUITE_NAME, PRAGMA, ...)                                         \
+	ztest_test_suite(SUITE_NAME, __VA_ARGS__);                                                 \
+	static struct unit_test_node z_ztest_test_node_##SUITE_NAME = {                            \
+		.name = #SUITE_NAME,                                                               \
+		.suite = _##SUITE_NAME,                                                            \
+		.pragma = PRAGMA,                                                                  \
+	};                                                                                         \
+	static int z_ztest_suite_##SUITE_NAME##_init(const struct device *dev)                     \
+	{                                                                                          \
+		ARG_UNUSED(dev);                                                                   \
+		z_ztest_test_node_##SUITE_NAME.next = z_ztest_unit_test_head;                      \
+		z_ztest_unit_test_head = &z_ztest_test_node_##SUITE_NAME;                          \
+		return 0;                                                                          \
+	}                                                                                          \
+	SYS_INIT(z_ztest_suite_##SUITE_NAME##_init, APPLICATION, 99)
```

Comment:
can we use a linker section instead of init+linked list?

Reply:
Sure I was planning on doing this in another PR to keep this one a bit simpler, but here's what I was thinking. @stephanosio could you take a look and let me know what you think:

1. Each node will also hold a `struct test_suite_stats` which will count `num_runs`, `num_failed`, and `num_skipped`.
2. Nodes will no longer be removed, but instead skipped if `num_runs != 0 && pragma == NULL`
3. Introduce a new macro `ztest_num_tests_not_run()` which will count the number of tests that have `num_runs == 0`.
4. Introduce `ztest_verify_all_tests_ran()` which will add a new test that simply checks that no node exists with `num_runs == 0`. This can be added at the end of `test_main()` to make sure that all the test suites actually did run.
5. Introduce `CONFIG_ZTEST_MAIN` that will add the following code to `ztest.c`:
```
#ifdef CONFIG_ZTEST_MAIN
void test_main(void)
{
  ztest_run_registered_test_suites();
  ztest_verify_all_tests_ran();
}
#endif /* CONFIG_ZTEST_MAIN */
```

I could also just add the `num_runs` component of the first bullet and implement the second in this PR, then leave the rest to the following. Please let me know if you would like me to leave things as they are, make minimal changes to this PR, or make all the changes in this PR.